pkg/virtctl/create/vm.go

```diff
@@ -0,0 +1,242 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2022 Red Hat, Inc.
+ *
+ */
+
+package create
+
+import (
+	"github.com/spf13/cobra"
+	k8sv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/api/instancetype"
+	"sigs.k8s.io/yaml"
+
+	"kubevirt.io/kubevirt/pkg/virtctl/templates"
+)
+
+const (
+	VM = "vm"
+
+	NameFlag                   = "name"
+	RunStrategyFlag            = "run-strategy"
+	InstancetypeFlag           = "instancetype"
+	ClusterInstancetypeFlag    = "cluster-instancetype"
+	PreferenceFlag             = "preference"
+	ClusterPreferenceFlag      = "cluster-preference"
+	PvcFlag                    = "pvc"
+	DatavolumeFlag             = "datavolume"
+	TerminationGracePeriodFlag = "termination-grace-period"
+)
+
+type createVM struct {
+	name                   string
+	instancetype           string
+	preference             string
+	pvc                    string
+	dataVolume             string
+	runStrategy            string
+	terminationGracePeriod int64
+}
+
+type optionFn func(*createVM, *v1.VirtualMachine)
+
+var optFns = map[string]optionFn{
+	InstancetypeFlag:        withInstancetype,
+	ClusterInstancetypeFlag: withClusterInstancetype,
+	PreferenceFlag:          withPreference,
+	ClusterPreferenceFlag:   withClusterPreference,
+	PvcFlag:                 withPVC,
+	DatavolumeFlag:          withDataVolume,
+}
+
+func NewVirtualMachineCommand() *cobra.Command {
+	c := defaultCreateVM()
+	cmd := &cobra.Command{
+		Use:     VM,
+		Short:   "Create a VirtualMachine manifest.",
+		Example: c.usage(),
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			return c.run(cmd)
+		},
+	}
+
+	cmd.Flags().StringVar(&c.name, NameFlag, c.name, "Specify the name of the VirtualMachine.")
+
+	cmd.Flags().StringVar(&c.instancetype, InstancetypeFlag, c.instancetype, "Specify the Instancetype of the VirtualMachine.")
+	cmd.Flags().StringVar(&c.instancetype, ClusterInstancetypeFlag, c.instancetype, "Specify the ClusterInstancetype of the VirtualMachine.")
+	cmd.MarkFlagsMutuallyExclusive(InstancetypeFlag, ClusterInstancetypeFlag)
+
+	cmd.Flags().StringVar(&c.preference, PreferenceFlag, c.preference, "Specify the Preference of the VirtualMachine.")
+	cmd.Flags().StringVar(&c.preference, ClusterPreferenceFlag, c.preference, "Specify the ClusterPreference of the VirtualMachine.")
+	cmd.MarkFlagsMutuallyExclusive(PreferenceFlag, ClusterPreferenceFlag)
+
+	cmd.Flags().StringVar(&c.pvc, PvcFlag, c.pvc, "Specify the PVC of the VirtualMachine.")
+	cmd.Flags().StringVar(&c.dataVolume, DatavolumeFlag, c.dataVolume, "Specify the DataVolume of the VirtualMachine.")
+	cmd.MarkFlagsMutuallyExclusive(PvcFlag, DatavolumeFlag)
```

Comment:
Users typically don't specify the PVC or DataVolume they want, instead they specify the source image they want to use for the VM which then gets imported into a pvc using the DataVolume flow.

Rather than a --pvc or --datavolume flag, i think a `--image-src` flag would make more sense. If the image source starts with an `http://|https://` then it's a http import, if it starts with a `docker://` then its a container disk import... We could come up with other options as well like perhaps a `pvc://` prefix for pvc cloning

Reply:
I added a `--image-source` flag but it works a bit differently than you described. I think we should discuss this a bit further.

Reply:
Just as a counter point to David's initial suggestion, we already have a command to orchestrate the upload of an image into the cluster, IMHO this command just needs to handle the creation of the `VirtualMachine` itself assuming the image is already present as a `DataSource` and/or `PVC`.